packages/core/core/src/PackagerRunner.js

```diff
@@ -430,52 +405,108 @@ export default class PackagerRunner {
           ).mode,
         };
+    let cacheKey = info.cacheKey;
+    let contentKey = getContentKey(cacheKey);
+    let contentStream = this.options.cache.getStream(contentKey);
+    let size = await writeFileStream(
+      outputFS,
+      filePath,
+      contentStream,
+      info.hashReferences,
+      hashRefToNameHash,
+      writeOptions,
+    );
+    bundle.stats = {
+      size,
+      time: info.time,
+    };
 
-    let size;
-    if (contents instanceof Readable) {
-      size = await writeFileStream(outputFS, filePath, contents, writeOptions);
-    } else {
-      await outputFS.writeFile(filePath, contents, writeOptions);
-      size = contents.length;
-    }
-
-    if (map != null) {
-      if (map instanceof Readable) {
-        await writeFileStream(outputFS, filePath + '.map', map);
-      } else {
-        await outputFS.writeFile(filePath + '.map', map);
-      }
+    let mapKey = getMapKey(cacheKey);
+    if (await this.options.cache.blobExists(mapKey)) {
+      let mapStream = this.options.cache.getStream(mapKey);
+      await writeFileStream(
+        outputFS,
+        filePath + '.map',
+        mapStream,
+        info.hashReferences,
+        hashRefToNameHash,
+      );
     }
-
-    return {size};
   }
 
   async writeToCache(cacheKey: string, contents: Blob, map: ?Blob) {
     let contentKey = getContentKey(cacheKey);
 
-    await this.options.cache.setStream(contentKey, blobToStream(contents));
+    let size = 0;
+    let hash = crypto.createHash('md5');
+    let prevBuf = '';
+    let hashReferences = [];
+    await this.options.cache.setStream(
+      contentKey,
+      blobToStream(contents).pipe(
+        new TapStream(buf => {
+          let str = prevBuf.toString() + buf.toString();
+          hashReferences = hashReferences.concat(
+            [...str.matchAll(/@@HASH_REFERENCE_\w{8}/g)].map(match => match[0]),
+          );
+          size += buf.length;
+          hash.update(buf);
+          prevBuf = buf;
+        }),
+      ),
+    );
 
     if (map != null) {
       let mapKey = getMapKey(cacheKey);
       await this.options.cache.setStream(mapKey, blobToStream(map));
     }
+
+    let info = {size, hash: hash.digest('hex'), hashReferences};
+    await this.options.cache.set(getInfoKey(cacheKey), JSON.stringify(info));
+    return info;
   }
 }
 
-function writeFileStream(
+async function writeFileStream(
   fs: FileSystem,
   filePath: FilePath,
   stream: Readable,
+  hashReferences: Array<string>,
+  hashRefToNameHash: Map<string, string>,
   options: ?FileOptions,
 ): Promise<number> {
-  return new Promise((resolve, reject) => {
-    let fsStream = fs.createWriteStream(filePath, options);
-    stream
-      .pipe(fsStream)
-      // $FlowFixMe
-      .on('finish', () => resolve(fsStream.bytesWritten))
-      .on('error', reject);
+  // TODO: stream directly to filesystem instead of to string first
+  //  - need to figure out how to replace hash references in a stream
```

Comment:
TODO: Skip stream replacement if we know there are no references to replace. We could probably start streaming them to dist immediately instead of waiting for all of the bundles to be processed to create the `hashRefToNameHash` mapping as well.

Reply:
> Skip stream replacement if we know there are no references to replace.

Done.

> We could probably start streaming them to dist immediately instead of waiting for all of the bundles to be processed to create the hashRefToNameHash mapping as well.

Not done yet.

Reply:
> We could probably start streaming them to dist immediately instead of waiting for all of the bundles to be processed to create the hashRefToNameHash mapping as well.

Done.